diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6fbd90c..5a131e9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,11 +12,13 @@ macro(add_target_property _target _prop _value)
 endmacro(add_target_property)
 
 if (CMAKE_VERSION VERSION_LESS "3.1")
-if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-  set (CMAKE_CXX_FLAGS "--std=gnu++11 ${CMAKE_CXX_FLAGS}")
-endif ()
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    set (CMAKE_CXX_FLAGS "--std=gnu++11 ${CMAKE_CXX_FLAGS}")
+  endif ()
 else ()
-  set (CMAKE_CXX_STANDARD 11)
+  if(NOT CMAKE_CXX_STANDARD)
+    set (CMAKE_CXX_STANDARD 11)
+  endif()
 endif ()
 
 list(APPEND CMAKE_PREFIX_PATH $ENV{ROOTSYS} ${ROOTSYS})
