diff --git a/CMakeLists.txt b/CMakeLists.txt
index 60ab7e6..b70ab36 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -52,7 +52,11 @@ if(NOT DEFINED NICAFEMTO_SUBDIR_BUILD)
 	If(NOT _HAS_CXX11_FLAG)
 	  Message(FATAL_ERROR "The used C++ compiler (${CMAKE_CXX_COMPILER}) does not support C++11. NicaFemto can only be compiled with compilers supporting C++11. Please install such an compiler.")
 	EndIf()
-	
+
+        if(APPLE)
+          SET(CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS "${CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS} -single_module -undefined dynamic_lookup")
+        endif()
+
         Execute_process(COMMAND $ENV{SIMPATH}/bin/fairsoft-config --cxxflags OUTPUT_VARIABLE _res_fairsoft_config OUTPUT_STRIP_TRAILING_WHITESPACE)
         String(FIND ${_res_fairsoft_config} "-std=c++11" POS_C++11)
         If(${POS_C++11} EQUAL -1)
@@ -131,7 +135,7 @@ if(NOT DEFINED NICAFEMTO_SUBDIR_BUILD)
           if(APPLE)
             set(GSL_DIR /usr/local)
           else()
-            unset(GSL_DIR)
+            set(GSL_DIR /usr)
           endif()
           find_package(GSL REQUIRED)
         endif()
diff --git a/cuts/CMakeLists.txt b/cuts/CMakeLists.txt
index 1d1c4f0..e5676ba 100755
--- a/cuts/CMakeLists.txt
+++ b/cuts/CMakeLists.txt
@@ -114,7 +114,7 @@ cutmonitors/NicaPropertyMonitorXYZ.cxx
 Set(HEADERS )
 Set(LINKDEF NicaCutLinkDef.h)
 Set(LIBRARY_NAME NicaCut)
-Set(DEPENDENCIES NicaFeatures)
+Set(DEPENDENCIES NicaFeatures NicaDataFormat Base)
 
 GENERATE_LIBRARY()
 
diff --git a/dataformat/CMakeLists.txt b/dataformat/CMakeLists.txt
index b8576ab..b3fc335 100644
--- a/dataformat/CMakeLists.txt
+++ b/dataformat/CMakeLists.txt
@@ -57,7 +57,7 @@ hiddeninfo/NicaV0Track.cxx
 Set(HEADERS )
 Set(LINKDEF NicaDataFormatLinkDef.h)
 Set(LIBRARY_NAME NicaDataFormat)
-Set(DEPENDENCIES NicaFeatures ${NICAFORMATLIBS})
+Set(DEPENDENCIES NicaFeatures ${NICAFORMATLIBS} Base EG)
 
 GENERATE_LIBRARY()
 
diff --git a/features/NicaHelixBase.cxx b/features/NicaHelixBase.cxx
index e248f54..7317950 100644
--- a/features/NicaHelixBase.cxx
+++ b/features/NicaHelixBase.cxx
@@ -8,6 +8,7 @@
  */
 #include "NicaHelixBase.h"
 #include <TMath.h>
+#include <TString.h>
 #include <iostream>

 Double_t NicaHelixBase::fgHelixBz = 0.5;
